### PR TITLE
Implement data generator for odin

### PR DESCRIPTION
**Pull request by tbukovics**
### Issue reference / description

The branch you merge from should already reference an event-formation-unit github ticket number. You can add a descriptive title, but if an issue is referenced, you don't have to.

## Checklist for submitter

- [x] Check for conflict with integration test
- [x] Unit tests pass

---

### Nominate for Group Code Review

- [x] Nominate for code review

### CODE REVIEW

File: src/modules/timepix3/generators/odin_datagen.bash [hunk 1]
```diff
@@ -0,0 +1,57 @@
+#!/bin/bash
+
+# Parse command line arguments
+while getopts "f:d" opt; do
```
Comment (mortenjc) [inline src/modules/timepix3/generators/odin_datagen.bash:4]: This is pretty old school :-) Are you familiar with argparse?

import argparse

https://docs.python.org/3/library/argparse.html

It is a bit more flexible IMO

File: src/modules/timepix3/generators/odin_datagen.bash [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+if [ ! -d "timepix_gen" ]; then
+    # Create a new virtual environment
+    python -m venv timepix_gen > /dev/null 2>&1
+    source timepix_gen/bin/activate
+    echo "Install necessary packages..."
+    pip install -r requirements.txt > /dev/null 2>&1
+fi
+
+# Define a function to be executed on exit
+cleanup() {
```
Comment (mortenjc) [inline src/modules/timepix3/generators/odin_datagen.bash:33]: I would add

function cleanup() {
}

File: src/modules/timepix3/generators/tpx_generator.py [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+PIXEL_TOA_OFFSET = 28
+PIXEL_TOT_OFFSET = 20
+PIXEL_FTOA_OFFSET = 16
+
+# TDC data constants
+TDC_TYPE_MASK = 0x0F00000000000000
+TDC_TRIGGERCOUNTER_MASK = 0x00FFF00000000000
+TDC_TIMESTAMP_MASK = 0x00000FFFFFFFFE00
+TDC_STAMP_MASK = 0x00000000000001E0
+TDC_TYPE_OFFSET = 56
```
Comment (mortenjc) [inline src/modules/timepix3/generators/tpx_generator.py:35]:
```
When dealing with masks I definitely prefer to align the fields 
TDC_TRIGGERCOUNTER_MASK =  0x00FFF00000000000
TDC_TIMESTAMP_MASK =       0x00000FFFFFFFFE00
TDC_STAMP_MASK =           0x00000000000001E0
```

**Review by mortenjc — APPROVED** [review]

File: src/modules/timepix3/generators/tpx_generator.py [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+
+                    types = (chunk & TYPE_MASK) >> TYPE_OFFSET
+                    tdc_mask = types == 0x6
+                    tdc_data = chunk[tdc_mask]
+
+                    tpx_mask = (types == 0x6) | (types == 0xB)
+
+                    data = chunk[tpx_mask]
+
+                    packet = Ether() / IP(dst="10.103.2.33") / UDP(sport=4096, dport=9888) / data.tobytes()
```
[the remark this answers is not in the document]
Comment (tbukovics) [reply src/modules/timepix3/generators/tpx_generator.py:112]: agree

File: src/modules/timepix3/generators/tpx_generator.py [hunk 5]
```diff
@@ -0,0 +1,240 @@
...
+        self.dtype = np.dtype(np.uint64)
+        self.udp_max_size = 8952
+        self.chunk_size = self.udp_max_size // self.dtype.itemsize
+        self.previous_time_ns = self.current_time_ns = time.time_ns()
+        self.frequency_hz = 10
+        self.period_ns = int(1 / self.frequency_hz * 1e9)
+        self.send_process = None
+        self.packet_time = time.time()
+        self.udp_dport = 9888
+        self.udp_sport = 4096
```
[the remark this answers is not in the document]
Comment (tbukovics) [reply src/modules/timepix3/generators/tpx_generator.py:76]: aggree

**Comment by tbukovics** [comment]
@morten and @amues 

I requested a new review, since I reworked a bit the python file. I also removed the bash, as we discussed with @amues this python will be delivered with EFU generators and the python env will be configured from ansible.

I modified the CMAKE to have a new script which can copy over a python file structure into the defined runtime folder, and set executeable permissions. So the python file should be moved to the build/generators by CMAKE

**Review by mortenjc — APPROVED** [review]

**Review by amues — APPROVED** [review]

**Pull request merged**